File: node/core/drench.js
const colors = require ('colors')

const drench = ( size = 6, maxCount = 20, matrix = [] ) => {

    let _matrix
    let game_over = false
    let _maxCount = maxCount
    let paintsCount = 0

    const init = () => {
        if ( matrix.length == 0 ) {
            _matrix = Array(size)
            for ( let i = 0; i < size; i++ ) {
                _matrix[i] = Array(size)
                for ( let j = 0; j < size; j++ ) {
                    _matrix[i][j] = randInt() % size
                }
            }
        }
    }

    const propagate = (color, oldColor, x = 0, y = 0  ) => {
        if ( x >= size || y >= size || x < 0 || y < 0 ) return
        if ( _matrix[x][y] != oldColor && ( x != 0 || y != 0 )) return
        _matrix[x][y] = color
        propagate ( color, oldColor, x, y + 1 )
        propagate ( color, oldColor, x, y - 1 )
        propagate ( color, oldColor, x + 1, y )
        propagate ( color, oldColor, x - 1, y )
    }

    const gameOver = () => {
        if ( paintsCount >= _maxCount ) 
            return true
        let val = _matrix[0][0]
        let cont = 0
        for ( let i = 0; i < size; i ++ ) {
            for ( let j = 0; j < size; j ++ ) {
                if ( _matrix[i][j] != val )
                    return false
                cont += 1
            }
        }
        return cont == ( size * size )
    }

    const remainingPaints = () => {
        return _maxCount - paintsCount
    }

    const print = () => {
        for ( let i = 0; i < size; i ++ ) {
            let str = ''
            for ( let j = 0; j < size; j ++ ) {
                str += array_colors[ _matrix[i][j] ] + ' '
            }
            console.log(str)
        }
        console.log()
    }

    const getOptions = () => {
        let aux = ''
        for ( let i = 0; i < size; i ++ ) 
            aux += array_colors[ i ] + ' '
        aux += '\n'
        for ( let i = 0; i < size; i ++ ) 
            aux += i + ' '
        aux += '\n'
        return aux
    }

    const updateMatrix = ( color ) => {
        if ( game_over || color == _matrix[0][0]) return
        paintsCount ++
        propagate ( color, _matrix[0][0] )
    }

    const randInt = ( max = 100, min = 20) => {
        return Math.floor(Math.random() * ( max - min + 1)) + min
    }

    const array_colors = [
        '▄'.green,
        '▄'.red,
        '▄'.blue,
        '▄'.yellow,
        '▄'.magenta,
        '▄'.cyan,
        '▄'.brightRed,
        '▄'.grey,
    ]

    return {
        'init'            : init,
        'getOptions'      : getOptions,
        'print'           : print,
        'remainingPaints' : remainingPaints,
        'isOver'          : gameOver,
        'propagateColor'  : propagate,
        'updateMatrix'    : updateMatrix
    }

}

module.exports = drench

File: python/core/drench.py
from random import randrange

class Drench :

    def __init__ ( self, size = 6, max_count = 20, matrix = [] ) :
        self.size = size
        self.game_over = False
        self.max_count = max_count
        self.paints_count = 0

        if len ( matrix ) == 0 :
            self.matrix = [[ randrange(self.size) for _ in range ( self.size )] for _ in range ( self.size )]
        else :
            self.matrix = matrix

    def get_options ( self ) :
        aux = ''
        for i in range ( self.size ) :
            aux += ( colors[i] + '▄' if i < len ( colors ) else '' ) + ' '
        aux += '\n'
        for i in range ( self.size ) :
            aux += str(i) + ' '
        return aux + '\n'

    def __repr__ ( self ) :
        aux = ''
        for i in range ( self.size ):
            for j in range ( self.size ):
                aux += (( colors[self.matrix[i][j]] + '▄' ) \
                    if i < self.size and \
                       j < self.size and \
                       self.matrix[i][j] < len(colors) \
                    else '' ) + ' '
            aux += '\n'
        return aux

    def remaining_paints ( self ) :
        return self.max_count - self.paints_count

    def is_over ( self ) :
        if self.paints_count >= self.max_count :
            return True
        val = self.matrix[0][0]
        cont = 0
        for i in range ( self.size ) :
            for j in range ( self.size ) :
                if self.matrix[i][j] != val  :
                    return False
                cont += 1
        return cont == ( self.size ** 2 )

    def _propagate_color ( self, color, old_color, x = 0, y = 0 ) :
        if x >= self.size or y >= self.size or x < 0 or y < 0 :
            return
        if self.matrix[x][y] != old_color and ( x != 0 or y != 0 ) :
            return
        self.matrix[x][y] = color

        self._propagate_color ( color, old_color, x, y + 1 )
        self._propagate_color ( color, old_color, x, y - 1 )
        self._propagate_color ( color, old_color, x + 1, y )
        self._propagate_color ( color, old_color, x - 1, y )

    def update_matrix ( self, color ) :
        if color >= self.size :
            raise Exception ('Color not supported')
        if self.game_over or color == self.matrix[0][0] :
            return
        self.paints_count += 1
        self._propagate_color ( color, self.matrix[0][0] )
        


        


colors = [

    '\33[31m',
    '\33[32m',
    '\33[33m',
    '\33[35m',
    '\33[95m',
    '\33[34m',
    '\33[94m',
    '\33[36m',
    '\33[105m',
    '\33[92m'

]